File: 백준/Gold/1759. 암호 만들기/암호 만들기.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [L, C] = input.shift().split(" ").map(Number);
const alphabet = input[0].split(" ").sort();
const password = [];
const vowel = ["a", "e", "i", "o", "u"];

function dfs(depth, start) {
  if (depth === L) {
    // 모음 갯수 카운트
    let cnt = 0;
    password.forEach((v) => {
      if (vowel.includes(v)) {
        cnt++;
      }
    });

    if (cnt >= 1 && L - cnt >= 2) console.log(password.join(""));
    return;
  }

  for (let i = start; i < C; i++) {
    if (!password.includes(alphabet[i])) {
      password.push(alphabet[i]);
      start++;
      dfs(depth + 1, start);
      password.pop();
    }
  }
}

dfs(0, 0);

File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
vowel = "aeiou"
L, C = map(int, input().split())
spells = list(input().split())
visited = [0] * C
stack = []
spells.sort()


def dfs(depth, start):
    if depth == L:
        cnt = 0
        for el in vowel:
            if el in stack:
                cnt += 1

        if cnt >= 1 and L - cnt >= 2:
            print(*stack, sep="")
        return

    for i in range(start, C):
        if not visited[i]:
            visited[i] = 1
            stack.append(spells[i])
            dfs(depth + 1, i)
            stack.pop()
            visited[i] = 0


dfs(0, 0)